alphagov/cf-terraform
# [#102924550] Normalise provisioning scripts between GCE and AWS

[Normalise provisioning scripts between GCE and AWS](https://www.pivotaltracker.com/story/show/102924550)

**What**

We currently have some discrepancies between GCE and AWS provisioning scripts. This story is to make these as consistent as possible. The pattern we intend to use is the current AWS pattern.

**How this PR should be reviewed**

This PR has been written to be reviewed in the following context:
- I want to move the setting of file permissions out of the provisioner script and into terraform and deprecate the use of `ssh-agent`, because it makes sense to have:
  -  The creation of a file, it's ownership and mode handled in the same location the [file provisioner](https://terraform.io/docs/provisioners/file.html) does not support the change of file permissions by [design and never will](https://github.com/hashicorp/terraform/issues/492) so we have used a [remote-exec](https://terraform.io/docs/provisioners/remote-exec.html).
  -  `ssh-agent` is only needed if we are not using the default location of the `id_rsa` private key.
- I want to create a new shared `provision.sh` script
- I want to install some base packages
- I do not want to redeploy bosh if it is already available
- I want to add details for the cf-release for the spiff templates and to optionally
  upload the release from local.
- I want to upload stemcells and releases
- I want to compile and deploy CF manifest using provision.sh
- Look Ma it's magic! I want to convert the yaml file `terraform_outputs*.yml` to a set of
  variables `terraform_output_*`, so we can use them in the provision script.
- I want to copy the logic from the original `gce/provision.sh` to setup the
  static IP in the BOSH command, only on GCE.
- I want to copy the scripts directory to bastion host
- I want to easily parse the bosh output to list releases and stemcells
- I want to fix the file paths and return codes in commands and lower the timeouts for faster feedback
- I want to restore the copying of account.json when deploying the bastion host
- I want to move the delete-route script to the scripts folder
- I want to check if stemcells and/or releases already exist, if so then skip over installing them
- I want to simply the logic in fix IP script
- I want to enable the new provision.sh script for gce and aws
- I want to set the correct path to /tmp/microbosh-info.json
- I want to ensure that the fix ip routing in GCE rule is defined
- I want to set the manifest file to be bosh-manifest.yml for destroy since I changed it earlier
- I need to correct location of manifest and script files since on a clean build these files are no longer present at the previously named locations
- I want to add an ingress rule to allow bastion to access bosh on it's external IP
  address (i.e. `AWS` Elastic IP)
- I want to Removing test for empty .bosh_config file
  If there an error logging into the bosh director an empty hash is inserted into the `.bosh_config` file. So if there is ever a failure you will never been able to log into the
  director unless the `.bosh_config` file is deleted.

**How to test this PR**

This PR can be tested by starting up a new environment in AWS and GCE, testing and then destroying it, followed by checking the relevant IaaS console to ensure all resources have been clean up correctly e.g.

For AWS:

```
make aws DEPLOY_ENV=$MY_ENVIRONMENT
make test-aws DEPLOY_ENV=$MY_ENVIRONMENT
make destroy-aws DEPLOY_ENV=$MY_ENVIRONMENT
```

For GCE

```
make gce DEPLOY_ENV=$MY_ENVIRONMENT
make test-gce DEPLOY_ENV=$MY_ENVIRONMENT
make destroy-gce DEPLOY_ENV=$MY_ENVIRONMENT
```

**Who should review this PR**

A Rumbustious Rhinoceros, if a rowdy single horned mammal is unavailable then a suitably fiesty member of the core team can serve as a substitute reviewer, it is optional if they have a horn or not.

Any member of the core team can merge this PR with the exception of @jimconner and @keymon who I paired with :p
